test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,126 @@
+	catalogSource         = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE", "test-catalog")
+	catalogSourcNamespace = constants.GetEnvWithDefault("SUBSCRIPTION_CATALOG_SOURCE_NAMESPACE", tutils.Namespace)
+	subPackage            = constants.GetEnvWithDefault("SUBSCRIPTION_PACKAGE", "infinispan")
+	subStartingCsv        = constants.GetEnvWithDefault("SUBSCRIPTION_STARTING_CSV", "infinispan-operator.v2.1.5")
+
+	latestImage = constants.GetEnvWithDefault("LATEST_OPERAND_IMAGE", constants.DefaultOperandImageOpenJDK)
+)
+
+func TestUpgrade(t *testing.T) {
+	testKube.NewNamespace(tutils.Namespace)
+
+	sub := &coreos.Subscription{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: coreos.SubscriptionCRDAPIVersion,
+			Kind:       coreos.SubscriptionKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      subName,
+			Namespace: subNamespace,
+		},
+		Spec: &coreos.SubscriptionSpec{
+			Channel:                sourceChannel,
+			CatalogSource:          catalogSource,
+			CatalogSourceNamespace: catalogSourcNamespace,
+			InstallPlanApproval:    coreos.ApprovalManual,
+			Package:                subPackage,
+			StartingCSV:            subStartingCsv,
+		},
+	}
+	testKube.CreateSubscription(sub)
+	defer testKube.DeleteNamespace(tutils.Namespace)
+
+	testKube.WaitForSubscriptionState(subName, subNamespace, coreos.SubscriptionStateUpgradePending)
+
+	// Upgrade the Subscription channel if required
+	if sourceChannel != latestChannel {
+		testKube.UpdateSubscriptionChannel(subName, subNamespace, latestChannel)
+	}
+
+	testKube.WaitForSubscriptionState(subName, subNamespace, coreos.SubscriptionStateUpgradePending)
+	// Approve the initial startingCSV InstallPlan
+	testKube.ApproveInstallPlan(subName, subNamespace, 1)
+
+	testKube.WaitForCrd(&apiextv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "infinispans.infinispan.org",
+		},
+	})
+
+	spec := tutils.DefaultSpec(testKube)
+	name := strcase.ToKebab(t.Name())
+	spec.Name = name
+	testKube.CreateInfinispan(spec, tutils.Namespace)
+	testKube.WaitForInfinispanPods(1, tutils.SinglePodTimeout, spec.Name, tutils.Namespace)
+	testKube.WaitForInfinispanCondition(spec.Name, spec.Namespace, ispnv1.ConditionWellFormed)
+
+	// Approve the upgrade InstallPlan
+	testKube.ApproveInstallPlan(subName, subNamespace, 2)
+	testKube.WaitForSubscriptionState(subName, subNamespace, coreos.SubscriptionStateAtLatest)
+
+	for _, state := range []ispnv1.ConditionType{ispnv1.ConditionUpgrade, ispnv1.ConditionStopping, ispnv1.ConditionWellFormed} {
+		testKube.WaitForInfinispanCondition(strcase.ToKebab(t.Name()), tutils.Namespace, state)
+	}
+
+	// Validates that all pods are running with desired image
+	pods := &corev1.PodList{}
+	err := testKube.Kubernetes.ResourcesList(tutils.Namespace, controllers.PodLabels(spec.Name), pods, context.TODO())
+	tutils.ExpectNoError(err)
+	for _, pod := range pods.Items {
+		if pod.Spec.Containers[0].Image != latestImage {
```

Comment:
The test performs only single version upgrade and it's from the previously released channel. This will fail with 2.2.1 release as it will expect 2.2.1 related server image but the subscription will be at this point only on 2.2.0

Reply:
As the server image has no relation to the operator version, I'm not sure we can fully automate this.

The test in it's current form will only work with a single upgrade anyway, as it approves the `InstallPlan` instances outside of a loop.

If we update the test to perform multiple upgrades in a loop, then I think we have two options:

1. Only check `pod.Spec.Containers[0].Image` for the final upgrade
2. Require a ENV var that contains a csv of all expected images. CSV will have to take into account operator versions that don't change server version.

My preference would be for 1.

Reply:
The relation between Operator and Server is given by CSV which specifies a particular server image it's supposed to run so it can parsed for each currently installed CSV for verification. And if we want the tests to go through several upgrades we'll need to wait for cluster being upgraded properly before proceeding. Updating CSV before the new version of cluster is installed could result in unpredictable behavior.

Basically approve new CSV version, wait till it's installed, parse it for DEFAULT_IMAGE, wait for cluster being upgraded. (Or simply proceed if the server wasn't updated in particular release)

Reply:
Actually I just realized that using `constants.DefaultOperandImageOpenJDK` won't work nicely with downstream as we'd have to set additional evn that's already available in the CSV

Reply:
Thanks for explaining in such detail. I was definitely under thinking it with my initial approach.

I have updated so that the server image is extracted from the currently active CSV.